internal/glimpl/gl_js.go

```diff
@@ -6,345 +6,485 @@ import (
-	f.Ctx.Call("scissor", x, y, width, height)
+	f.Ctx.Call(_glScissor, uintptr(unsafe.Pointer(&x)), uintptr(unsafe.Pointer(&y)), uintptr(unsafe.Pointer(&width)), uintptr(unsafe.Pointer(&height)))
 }
 func (f *Functions) ShaderSource(s Shader, src string) {
-	f.Ctx.Call("shaderSource", js.Value(s), src)
+	f.Ctx.Call(_glShaderSource, uintptr(unsafe.Pointer(&s)), uintptr(unsafe.Pointer(&src)))
 }
 func (f *Functions) TexImage2D(target Enum, level int, internalFormat int, width, height int, format, ty Enum, data []byte) {
-	f.Ctx.Call("texImage2D", int(target), int(level), int(internalFormat), int(width), int(height), 0, int(format), int(ty), f.byteArrayOf(data))
+	f.Ctx.Call(_glTexImage2D, uintptr(unsafe.Pointer(&target)), uintptr(unsafe.Pointer(&level)), uintptr(unsafe.Pointer(&internalFormat)), uintptr(unsafe.Pointer(&width)), uintptr(unsafe.Pointer(&height)), uintptr(unsafe.Pointer(&_zero)), uintptr(unsafe.Pointer(&format)), uintptr(unsafe.Pointer(&ty)), uintptr(unsafe.Pointer(&data)))
 }
 func (f *Functions) TexSubImage2D(target Enum, level int, x, y, width, height int, format, ty Enum, data []byte) {
-	f.Ctx.Call("texSubImage2D", int(target), level, x, y, width, height, int(format), int(ty), f.byteArrayOf(data))
+	f.Ctx.Call(_glTexSubImage2D, uintptr(unsafe.Pointer(&target)), uintptr(unsafe.Pointer(&level)), uintptr(unsafe.Pointer(&x)), uintptr(unsafe.Pointer(&y)), uintptr(unsafe.Pointer(&width)), uintptr(unsafe.Pointer(&height)), uintptr(unsafe.Pointer(&format)), uintptr(unsafe.Pointer(&ty)), uintptr(unsafe.Pointer(&data)))
 }
 func (f *Functions) TexParameteri(target, pname Enum, param int) {
-	f.Ctx.Call("texParameteri", int(target), int(pname), int(param))
+	f.Ctx.Call(_glTexParameteri, uintptr(unsafe.Pointer(&target)), uintptr(unsafe.Pointer(&pname)), uintptr(unsafe.Pointer(&param)))
 }
 func (f *Functions) UniformBlockBinding(p Program, uniformBlockIndex uint, uniformBlockBinding uint) {
-	f.Ctx.Call("uniformBlockBinding", js.Value(p), int(uniformBlockIndex), int(uniformBlockBinding))
+	f.Ctx.Call(_glUniformBlockBinding, uintptr(unsafe.Pointer(&p)), uintptr(unsafe.Pointer(&uniformBlockIndex)), uintptr(unsafe.Pointer(&uniformBlockBinding)))
 }
-func (f *Functions) Uniform1f(dst Uniform, v float32) {
-	f.Ctx.Call("uniform1f", js.Value(dst), v)
+func (f *Functions) Uniform1f(dst Uniform, v0 float32) {
+	vv0 := float64(v0) // Float32 doesn't work
+	f.Ctx.Call(_glUniform1f, uintptr(unsafe.Pointer(&dst)), uintptr(unsafe.Pointer(&vv0)))
 }
 func (f *Functions) Uniform1i(dst Uniform, v int) {
-	f.Ctx.Call("uniform1i", js.Value(dst), v)
+	f.Ctx.Call(_glUniform1i, uintptr(unsafe.Pointer(&dst)), uintptr(unsafe.Pointer(&v)))
 }
 func (f *Functions) Uniform2f(dst Uniform, v0, v1 float32) {
-	f.Ctx.Call("uniform2f", js.Value(dst), v0, v1)
+	vv0, vv1 := float64(v0), float64(v1) // Float32 doesn't work
+	f.Ctx.Call(_glUniform2f, uintptr(unsafe.Pointer(&dst)), uintptr(unsafe.Pointer(&vv0)), uintptr(unsafe.Pointer(&vv1)))
```

Comment:
Uniform etc. are uints. Won't all these &value pointer taking expressions lead the values to be heap-allocated and thus subject to garbage collection?

Reply:
I think you are right, did you have any idea to avoid it? Using `[]interface` seems to have the same issue, actually even worse.

About the `float64()` it's also done internally by the `syscall/js`, https://github.com/golang/go/blob/master/src/syscall/js/js.go#L162-L221. I'm trying to read the `float32` directly on the JS too.

Reply:
See my comment about generating the assembly stubs.